cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -319,31 +323,26 @@ public void getAllNamespaceSourceControlMetadata(FullHttpRequest request, HttpRe
       @QueryParam("filter") String filter
   ) throws Exception {
     validateNamespace(namespaceId);
-
-    JsonPaginatedListResponder.respond(GSON, responder, APP_LIST_PAGINATED_KEY_SHORT,
-        jsonListResponder -> {
-          AtomicReference<SourceControlMetadataRecord> lastRecord = new AtomicReference<>(null);
-          ScanSourceControlMetadataRequest scanRequest = SourceControlMetadataHelper.getScmStatusScanRequest(
-              namespaceId,
-              pageToken, pageSize, sortOrder, sortOn, filter);
-          boolean pageLimitReached = false;
-          try {
-            pageLimitReached = applicationLifecycleService.scanSourceControlMetadata(
-                scanRequest, batchSize,
-                scmMetaRecord -> {
-                  jsonListResponder.send(scmMetaRecord);
-                  lastRecord.set(scmMetaRecord);
-                });
-          } catch (IOException e) {
-            responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
-          }
-          SourceControlMetadataRecord record = lastRecord.get();
-          return !pageLimitReached || record == null ? null : record.getName();
-        });
+    List<SourceControlMetadataRecord> apps = new ArrayList<>();
+    ScanSourceControlMetadataRequest scanRequest = SourceControlMetadataHelper.getScmStatusScanRequest(
+        namespaceId,
+        pageToken, pageSize, sortOrder, sortOn, filter);
+    boolean pageLimitReached;
+    pageLimitReached = applicationLifecycleService.scanSourceControlMetadata(
+        scanRequest, batchSize,
+        apps::add);
+    SourceControlMetadataRecord record = apps.isEmpty() ? null :apps.get(apps.size() - 1);
+    String nextPageToken = !pageLimitReached || record == null ? null :
+        record.getName();
+    long lastRefreshTime = applicationLifecycleService.getLastRefreshTime(namespaceId);
```

Comment:
Seems like unwise to sync all apps, shouldn't we just sync the apps fetched in this page?

Reply:
As we are cloning the full repository which would be time taking it might be wasteful to just update few pipelines.